New routing encodings, bundled constraints, design passes

makaimann opened the pull request:
Restructured netlist modifications into passes in design/passes.py

Renamed nets to ties. Nets now refer to sets of ties.
Note: This is implemented but currently unused. The net passes are not being called. Nets as they are now are not timing consistent, but we can modify that in the future if we ever use the net representation.
    
Updated routing encodings and deprecated old encodings. build_spnr takes a region value and allows monosat to replace components within a region of that length
    
Bundled routing settings into functions

Commit:
Works towards full support for feedthroughs

Commit:
Trying with multigraph encoding

Commit:
Works towards simultaneous place and route encoding

Commit:
Moving some of design building to passes

Commit:
Fusing registers now

Commit:
Adding new encodings and net/ties distinction

Commit:
Encoding AtMostOne based on number of edges

Commit:
Trying regional pnr

Bitstream writer does not support regions != 0 -- i.e. not recovering new placement

Commit:
Clean up, tune distance constraints and make constraint bundles

Restructured netlist modifications into passes in design/passes.py

Renamed nets to ties. Nets now refer to sets of ties.
Note: Currently nets are not being built, but the functionality is
there if we need it in the future

Updated routing encodings and deprecated old encodings.

Bundled routing settings into functions

makaimann commented:
Here's the CGRAFlow Travis build with these changes: https://travis-ci.org/makaimann/CGRAFlow/builds/263239602

cdonovick commented:
I am not clear on the purpose of module.collapse_input

Ideally I think passes should be passed to the constructor (like constraints are passed to place_design).  If they are not being passed and are not going to be used outside of the Design class, they should probably just be private methods or functions not included in `__all__`.  

Example of possible passing architecture:

```
# Stage 1: passes take (modules, ties) 
# Return modules, ties but pass may have changed types
(modules, ties) = _io_hack(modules, ties)
(modules, ties) = _build_modules(modules, ties) 
(modules, ties) = _build_ties(modules, ties)
# end Stage 1

self._modules=frozenset(modules.values())
self._ties=frozenset(ties.values())
p_modules, p_ties = modules, ties

# Stage 2: passes take p_modules, p_ties
(p_modules, p_ties) = fuse_comps(modules, ties)
# end Stage 2

self._p_modules = frozenset(_p_modules)
self._p_ties = frozenset(_p_ties)
```

I think one thing we should consider is replacing the p_modules, p_ties with a general idea of a view.  
So `design.view['physical'].modules -> design.physical_modules`
This is distinct from the modules_with_attr construct as it doesn't just filter the graph, a view could return an entirely different graph.

makaimann commented:
```module.collapse_input``` is for fusing modules. It can be used for arbitrary inputs (as long as those only have one input), but currently it's only used for fused registers. It returns the new tie (fused_module's input connected to self) after fusing the module.

That's a good point about passes. Either is probably fine. I think it will probably only be used in Design, so we could just make them private methods. On the other hand, that makes the Design class larger... I'll put it in Design unless you have objections. I don't think there's any reason for the user to see the passes -- at least not yet.

I like that view idea. It's also distinct from ```modules_with_attr``` because it operates on graphs/designs not just modules. So ```design.view['physical']``` would return another design, from which you can access modules or ties.

Commit:
Improving design passes structure based on Caleb's feedback

Adding comments

cdonovick commented:
I want to maintain the invariant that `for all t in ties: t in t.dst.inputs and t in t.src.outputs`

We can make fusion permanent and remove the concept of physical_modules / physical_ties  or
discuss the possibility of views further.

makaimann commented:
That sounds good. How about making fusion permanent for now? I don't see why we would need ties to fused modules anyway. But I definitely think we should discuss the views possibility still.

cdonovick commented:
Sounds good to me

Commit:
Updates register fusing

This commit makes it so all (Reg, PE/IO) pairs fuse

Removed collapse_inputs and build_nets

makaimann commented:
I made the changes we discussed. I also got rid of build_nets, collapse_input, etc... If we decide to use it later, we can always use git magic to bring it back. So we can re-rename ties back to nets if you prefer?

Registers are now fused to all non-register outputs (method 2 from your slack message). But the register's resource is only changed to Fused if **all** its outputs are PEs (for now, IOs are still PEs). It's a little strange if you think about it as modules, but I think the final result makes sense. Fused registers mean "entirely fused". Otherwise, they are fused into the output PEs which makes new ties, but the modules still stick around for connection to their register outputs. See the comments in ```design/design.py:fuse_reg```

There are only ```ties``` now, no ```physical_ties```. However, I left the distinction between ```modules``` and ```physical_modules```, because it might still be nice to loop through all modules. What do you think? Do you ever need to access Const/Fused modules?

cdonovick commented:
The only time we need modules vs physical modules is during bitstream writing.   If we can handle bitstream writing without the fused modules than we should probably remove them so we can maintain 
```
for m in modules:
    for t in m.outputs: 
      assert m is t.src
    for t in m.inputs:
      assert m is t.dst
```

We could also have fused modules have no inputs or outputs and just have a location which is equal to the module they are fused to.

cdonovick commented:
Adding to my previous comment, if we have fused modules have no inputs or outputs part of their location would probably have to be the port they are attached to.

cdonovick committed:
Preprocess fusion of register

cdonovick commented:
For posterity register fusion can be done in two ways:
Given design:
```
m1 -- r1 -- r2 -- m3
      |
      m2
```
Can be transformed to:

``` 
m1 -- r1 -- (r2 + m3)
      |
      m2
```
or
```
m1 -- r1 -- (r2 + m3)
|
(r1 + m2)
```

We have opted for option two as it is more flexible for routing

Commit:
Fuse preprocessed registers and keep set of registered ports in module

Deprecate physical_modules

cdonovick committed:
Update assertions
